src/seraphis_main/scan_balance_recovery_utils.cpp

```diff
@@ -423,41 +420,69 @@ bool try_find_legacy_enotes_in_tx(const rct::key &legacy_base_spend_pubkey,
+    // 4. compute the key derivations for all main enote ephemeral pubkeys
+    rct::key_keyV_variant temp_DH_derivations = rct::key{};
+    if (legacy_main_enote_ephemeral_pubkeys.is_type<rct::keyV>())
+    {
+        temp_DH_derivations = rct::keyV{};
+        temp_DH_derivations.unwrap<rct::keyV>().reserve(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::keyV>().size());
+        for (rct::key enote_ephemeral_pubkey : legacy_main_enote_ephemeral_pubkeys.unwrap<rct::keyV>()) {
+            hwdev.generate_key_derivation(rct::rct2pk(enote_ephemeral_pubkey), legacy_view_privkey, temp_DH_derivation);
+            temp_DH_derivations.unwrap<rct::keyV>().emplace_back((rct::key &) temp_DH_derivation);
+        }
+    }
+    else
+    {
+        hwdev.generate_key_derivation(rct::rct2pk(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::key>()), legacy_view_privkey, temp_DH_derivation);
+        temp_DH_derivations = (rct::key &) temp_DH_derivation;
+    }
 
-    // 6. scan all enotes using the main key derivation
+    // 5. scan all enotes using key derivations for every ephemeral pub key
     for (std::size_t enote_index{0}; enote_index < enotes_in_tx.size(); ++enote_index)
     {
-        // a. try to recover a contextual basic record from the enote
-        if (!try_view_scan_legacy_enote_v1(legacy_base_spend_pubkey,
-                legacy_subaddress_map,
-                block_index,
-                block_timestamp,
-                transaction_id,
-                total_enotes_before_tx,
-                enote_index,
-                unlock_time,
-                tx_memo,
-                enotes_in_tx[enote_index],
-                legacy_main_enote_ephemeral_pubkey,
-                temp_DH_derivation,
-                origin_status,
-                hwdev,
-                temp_contextual_record))
-            continue;
+        // all ephemeral pub key - DH_derivation pairs
+        for (std::size_t pk_index{0};
+                (temp_DH_derivations.is_type<rct::keyV>() && pk_index < temp_DH_derivations.unwrap<rct::keyV>().size()) ||
+                (temp_DH_derivations.is_type<rct::key>() && pk_index < 1);
+                ++pk_index)
+        {
+            crypto::public_key temp_legacy_main_enote_ephemeral_pubkey = legacy_main_enote_ephemeral_pubkeys.is_type<rct::key>() ?
+                        rct::rct2pk(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::key>()) :
+                        rct::rct2pk(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::keyV>()[pk_index]);
+            temp_DH_derivation = temp_DH_derivations.is_type<rct::key>() ?
+                        (crypto::key_derivation &) (temp_DH_derivations.unwrap<rct::key>()) :
+                        (crypto::key_derivation &) (temp_DH_derivations.unwrap<rct::keyV>()[pk_index]);
+            // a. try to recover a contextual basic record from the enote
+            if (!try_view_scan_legacy_enote_v1(legacy_base_spend_pubkey,
```

Comment:
```suggestion
            crypto::public_key temp_legacy_main_enote_ephemeral_pubkey = legacy_main_enote_ephemeral_pubkeys.is_type<rct::key>() ?
                        rct::rct2pk(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::key>()) :
                        rct::rct2pk(legacy_main_enote_ephemeral_pubkeys.unwrap<rct::keyV>()[pk_index]);
            temp_DH_derivation = temp_DH_derivations.is_type<rct::key>() ?
                        (crypto::key_derivation &) (temp_DH_derivations.unwrap<rct::key>()) :
                        (crypto::key_derivation &) (temp_DH_derivations.unwrap<rct::keyV>()[pk_index]);

            // a. try to recover a contextual basic record from the enote
            if (!try_view_scan_legacy_enote_v1(legacy_base_spend_pubkey,
```

I would add a helper function for accessing by index. You can add it as a static function to this file.